File: typescript/src/namespaces/streaming.ts
import type { Credentials } from "../auth.js";
import { buildHeaders } from "../auth.js";
import type { StreamEvent } from "../types.js";

/**
 * O(1) linked-list queue — replaces Array.shift() which is O(N).
 */
interface QueueNode<T> {
  value: T;
  next: QueueNode<T> | null;
}

class SimpleQueue<T> {
  private head: QueueNode<T> | null = null;
  private tail: QueueNode<T> | null = null;
  private _length = 0;

  get length() {
    return this._length;
  }

  enqueue(value: T) {
    const node: QueueNode<T> = { value, next: null };
    if (this.tail) this.tail.next = node;
    else this.head = node;
    this.tail = node;
    this._length++;
  }

  dequeue(): T | undefined {
    if (!this.head) return undefined;
    const value = this.head.value;
    this.head = this.head.next;
    if (!this.head) this.tail = null;
    this._length--;
    return value;
  }
}

/**
 * WebSocket event stream for real-time Council updates.
 */
export class EventStream {
  private ws: WebSocket | null = null;
  private subscriptions: string[] = [];
  private closed = false;
  private credentials: Credentials;

  constructor(credentials: Credentials) {
    this.credentials = credentials;
  }

  /**
   * Connect to the WebSocket event stream.
   */
  async connect(): Promise<void> {
    const base = this.credentials.baseUrl
      .replace("http://", "ws://")
      .replace("https://", "wss://");
    const url = `${base}/events`;

    // Use native WebSocket (browser/Node 21+) or ws package
    if (typeof WebSocket !== "undefined") {
      this.ws = new WebSocket(url);
    } else {
      try {
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const wsModule = await import("ws");
        const WS = wsModule.default ?? wsModule;
        // ws package constructor accepts options as second param
        this.ws = new (WS as any)(url, {
          headers: buildHeaders(this.credentials),
        }) as unknown as WebSocket;
      } catch {
        throw new Error(
          "WebSocket not available. Install the ws package: npm install ws",
        );
      }
    }

    return new Promise<void>((resolve, reject) => {
      if (!this.ws) return reject(new Error("WebSocket not initialized"));

      this.ws.onopen = () => {
        // Send auth
        this.ws?.send(
          JSON.stringify({
            type: "auth",
            token:
              this.credentials.accessToken ??
              this.credentials.jwtToken ??
              this.credentials.apiKey,
          }),
        );
        resolve();
      };

      this.ws.onerror = (err) => {
        reject(new Error(`WebSocket error: ${err}`));
      };
    });
  }

  /**
   * Subscribe to a channel (e.g., 'agent:agent_abc123', 'jury:*').
   */
  async subscribe(channel: string): Promise<void> {
    this.subscriptions.push(channel);
    if (this.ws && this.ws.readyState === 1) {
      this.ws.send(JSON.stringify({ type: "subscribe", channel }));
    }
  }

  /**
   * Unsubscribe from a channel.
   */
  async unsubscribe(channel: string): Promise<void> {
    this.subscriptions = this.subscriptions.filter((s) => s !== channel);
    if (this.ws && this.ws.readyState === 1) {
      this.ws.send(JSON.stringify({ type: "unsubscribe", channel }));
    }
  }

  /**
   * Close the WebSocket connection.
   */
  async close(): Promise<void> {
    this.closed = true;
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  /**
   * Async iterator for events.
   */
  async *[Symbol.asyncIterator](): AsyncIterableIterator<StreamEvent> {
    if (!this.ws) {
      await this.connect();
    }

    // Re-subscribe
    for (const channel of this.subscriptions) {
      this.ws?.send(JSON.stringify({ type: "subscribe", channel }));
    }

    const ws = this.ws!;
    const eventQueue = new SimpleQueue<StreamEvent>();
    let resolveWait: (() => void) | null = null;
    let done = false;

    ws.onmessage = (event: MessageEvent) => {
      try {
        const data = JSON.parse(String(event.data)) as Record<string, unknown>;
        const msgType = String(data.type ?? "");

        // Skip internal messages
        if (
          [
            "subscribe_ack",
            "unsubscribe_ack",
            "ping",
            "pong",
            "auth_success",
          ].includes(msgType)
        ) {
          return;
        }

        eventQueue.enqueue({
          type: msgType,
          data: (data.data ?? data) as Record<string, unknown>,
          timestamp: data.timestamp
            ? new Date(data.timestamp as string)
            : undefined,
        });

        if (resolveWait) {
          resolveWait();
          resolveWait = null;
        }
      } catch {
        // Ignore parse errors
      }
    };

    ws.onclose = () => {
      done = true;
      if (resolveWait) {
        resolveWait();
        resolveWait = null;
      }
    };

    while (!done && !this.closed) {
      if (eventQueue.length > 0) {
        yield eventQueue.dequeue()!;
      } else {
        await new Promise<void>((resolve) => {
          resolveWait = resolve;
        });
      }
    }

    // Drain remaining
    while (eventQueue.length > 0) {
      yield eventQueue.dequeue()!;
    }
  }
}

File: python/council-sdk/council/streaming.py
"""WebSocket streaming for real-time Council events."""

from __future__ import annotations

import asyncio
import json
from typing import Any, AsyncIterator, Optional

from council.auth import Credentials
from council.types import StreamEvent


class EventStream:
    """WebSocket event stream for real-time updates.

    Usage::

        async with client.stream() as stream:
            await stream.subscribe("agent:agent_abc123")
            await stream.subscribe("jury:*")

            async for event in stream:
                print(f"{event.type}: {event.data}")
    """

    def __init__(self, credentials: Credentials) -> None:
        self._credentials = credentials
        self._ws: Any = None
        self._subscriptions: list[str] = []
        self._closed = False

    async def connect(self) -> None:
        """Establish the WebSocket connection."""
        try:
            import websockets
        except ImportError:
            raise ImportError(
                "websockets is required for streaming. "
                "Install it with: pip install council-sdk"
            )

        base = self._credentials.base_url.replace("http://", "ws://").replace(
            "https://", "wss://"
        )
        url = f"{base}/events"

        headers = self._credentials.auth_header
        self._ws = await websockets.connect(url, additional_headers=headers)

    async def subscribe(self, channel: str) -> None:
        """Subscribe to a channel (e.g., 'agent:agent_abc123', 'jury:*')."""
        self._subscriptions.append(channel)
        if self._ws:
            await self._ws.send(
                json.dumps({"type": "subscribe", "channel": channel})
            )

    async def unsubscribe(self, channel: str) -> None:
        """Unsubscribe from a channel."""
        self._subscriptions = [s for s in self._subscriptions if s != channel]
        if self._ws:
            await self._ws.send(
                json.dumps({"type": "unsubscribe", "channel": channel})
            )

    async def close(self) -> None:
        """Close the WebSocket connection."""
        self._closed = True
        if self._ws:
            await self._ws.close()
            self._ws = None

    def __aiter__(self) -> AsyncIterator[StreamEvent]:
        return self._iter_events()

    async def _iter_events(self) -> AsyncIterator[StreamEvent]:
        """Iterate over incoming events."""
        if not self._ws:
            await self.connect()

        assert self._ws is not None

        # Re-subscribe after connect
        for channel in self._subscriptions:
            await self._ws.send(
                json.dumps({"type": "subscribe", "channel": channel})
            )

        try:
            async for message in self._ws:
                if self._closed:
                    break

                try:
                    data = json.loads(message) if isinstance(message, str) else json.loads(message)
                except json.JSONDecodeError:
                    continue

                # Skip internal messages
                msg_type = data.get("type", "")
                if msg_type in ("subscribe_ack", "unsubscribe_ack", "ping", "pong"):
                    continue

                yield StreamEvent(
                    type=msg_type,
                    data=data.get("data", data),
                    timestamp=data.get("timestamp"),
                )
        except Exception:
            if not self._closed:
                raise

    async def __aenter__(self) -> "EventStream":
        await self.connect()
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.close()
